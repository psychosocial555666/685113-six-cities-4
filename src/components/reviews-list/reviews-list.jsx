import React, {PureComponent} from "react";
import {CommentType, MAX_COOMMENT_LENGTH} from "../../const";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import ReviewsItem from "../reviews-item/reviews-item.jsx";
import {getCurrentComments} from "../../reducer/data/selectors.js";
import {Operation as DataOperation} from "../../reducer/data/data.js";
import {adaptComments} from "../../adapter/comments.js";
import {getAuthorizationStatus} from "../../reducer/user/selectors.js";
import CommentForm from "../comment-form/comment-form.jsx";


class ReviewsList extends PureComponent {
  constructor(props) {
    super(props);

  }

  componentDidMount() {
    const {id, onLoadComments} = this.props;
    onLoadComments(id);
  }

  render() {
    const {comments, authorizationStatus, id} = this.props;
    const sortedComments = comments.slice(0, comments.length).sort((a, b) => new Date(b.reviewTime) - new Date(a.reviewTime)).slice(0, MAX_COOMMENT_LENGTH);

    return (
      <section className="property__reviews reviews">
        <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{comments.length}</span></h2>
        <ul className="reviews__list">
          {sortedComments.map((it) => <ReviewsItem
            key={it.reviewId}
            review={it}
          />)}
        </ul>
        {authorizationStatus ? <CommentForm id={id}/> : ``}
      </section>
    );
  }
}

ReviewsList.propTypes = {
  comments: CommentType,
  id: PropTypes.number.isRequired,
  onLoadComments: PropTypes.func.isRequired,
  authorizationStatus: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  comments: getCurrentComments(state),
  authorizationStatus: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onLoadComments(id) {
    dispatch(DataOperation.loadComments(adaptComments, id));
  },
});

export {ReviewsList};

export default connect(mapStateToProps, mapDispatchToProps)(ReviewsList);
