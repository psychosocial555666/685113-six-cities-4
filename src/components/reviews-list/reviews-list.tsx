import * as React from "react";
import {sortComments} from "../../utils";
import {connect} from "react-redux";
import ReviewsItem from "../reviews-item/reviews-item";
import {getCurrentComments} from "../../reducer/data/selectors";
import {Operation as DataOperation} from "../../reducer/data/data";
import {adaptComments} from "../../adapter/comments";
import {getAuthorizationStatus} from "../../reducer/user/selectors";
import CommentForm from "../comment-form/comment-form";
import {AuthorizationStatus} from "../../reducer/user/user";
import { Review } from "../../types";

interface Props {
  comments: Review[],
  id: number,
  onLoadComments: (id: number) => void,
  authorizationStatus: string,
};

class ReviewsList extends React.PureComponent<Props, {}> {
  constructor(props) {
    super(props);

  }

  componentDidMount() {
    const {id, onLoadComments} = this.props;
    onLoadComments(id);
  }

  render() {
    const {comments, authorizationStatus, id} = this.props;
    const sortedComments = sortComments(comments)

    return (
      <section className="property__reviews reviews">
        <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{comments.length}</span></h2>
        <ul className="reviews__list">
          {sortedComments.map((it) => <ReviewsItem
            key={it.reviewId}
            review={it}
          />)}
        </ul>
        {authorizationStatus === AuthorizationStatus.AUTH ? <CommentForm id={id}/> : ``}
      </section>
    );
  }
}

const mapStateToProps = (state) => ({
  comments: getCurrentComments(state),
  authorizationStatus: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onLoadComments(id) {
    dispatch(DataOperation.loadComments(adaptComments, id));
  },
});

export {ReviewsList};

export default connect(mapStateToProps, mapDispatchToProps)(ReviewsList);
