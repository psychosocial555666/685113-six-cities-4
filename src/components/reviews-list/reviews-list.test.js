import React from "react";
import renderer from "react-test-renderer";
import {ReviewsList} from "./reviews-list.jsx";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import NameSpace from "../../reducer/name-space.js";

const mockStore = configureStore([]);

const comments = [
  {
    reviewId: 14,
    authorId: 21,
    reviewName: `Ozzy`,
    reviewAvatar: `src`,
    reviewGrade: 4,
    reviewText: `Id minim labore ut velit sit velit.Magna deserunt reprehenderit consequat elit cupidatat proident nostrud amet minim nulla.`,
    reviewTime: `May 8, 2016`,
    isPro: false,
  },
];

const onLoadComments = jest.fn();

describe(`ReviewsList`, () => {
  it(`Should ReviewsList render correctly`, () => {
    const store = mockStore({
      [NameSpace.DATA]: {
        comments: [
          {
            reviewId: 14,
            authorId: 21,
            reviewName: `Ozzy`,
            reviewAvatar: `src`,
            reviewGrade: 4,
            reviewText: `Id minim labore ut velit sit velit.Magna deserunt reprehenderit consequat elit cupidatat proident nostrud amet minim nulla.`,
            reviewTime: `May 8, 2016`,
            isPro: false,
          },
        ],
      },
    });
    const tree = renderer
        .create(<Provider store={store}>
          <ReviewsList
            comments={comments}
            id={1}
            onLoadComments={onLoadComments} />
        </Provider>, {
          createNodeMock: () => {
            return document.createElement(`div`);
          }
        }
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

