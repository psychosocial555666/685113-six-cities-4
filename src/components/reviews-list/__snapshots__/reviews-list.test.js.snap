// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should OfferList render correctly 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews Â· 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="src"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Ozzy
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          Id minim labore ut velit sit velit.Magna deserunt reprehenderit consequat elit cupidatat proident nostrud amet minim nulla.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-04-24"
        >
          May 8, 2016
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="src"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Mick
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "40%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          Incididunt fugiat non aliqua eu nisi.Id Lorem cillum non voluptate nulla id fugiat Lorem exercitation irure ullamco enim veniam ullamco.
        </p>
        <time
          className="reviews__time"
          dateTime="2019-04-24"
        >
          June 5, 2018
        </time>
      </div>
    </li>
  </ul>
  <form
    action="#"
    className="reviews__form form"
    method="post"
  >
    <label
      className="reviews__label form__label"
      htmlFor="review"
    >
      Your review
    </label>
    <div
      className="reviews__rating-form form__rating"
    >
      <input
        className="form__rating-input visually-hidden"
        id="5-stars"
        name="rating"
        type="radio"
        value="5"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="5-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        className="form__rating-input visually-hidden"
        id="4-stars"
        name="rating"
        type="radio"
        value="4"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="4-stars"
        title="good"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        className="form__rating-input visually-hidden"
        id="3-stars"
        name="rating"
        type="radio"
        value="3"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="3-stars"
        title="not bad"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        className="form__rating-input visually-hidden"
        id="2-stars"
        name="rating"
        type="radio"
        value="2"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="2-stars"
        title="badly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        className="form__rating-input visually-hidden"
        id="1-star"
        name="rating"
        type="radio"
        value="1"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="1-star"
        title="terribly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
    </div>
    <textarea
      className="reviews__textarea form__textarea"
      id="review"
      name="review"
      placeholder="Tell how was your stay, what you like and what can be improved"
    />
    <div
      className="reviews__button-wrapper"
    >
      <p
        className="reviews__help"
      >
        To submit review please make sure to set 
        <span
          className="reviews__star"
        >
          rating
        </span>
         and describe your stay with at least 
        <b
          className="reviews__text-amount"
        >
          50 characters
        </b>
        .
      </p>
      <button
        className="reviews__submit form__submit button"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</section>
`;
